src/sema/mutability.rs

```diff
@@ -27,9 +27,15 @@ pub fn mutability(file_no: usize, ns: &mut Namespace) {
 /// While we recurse through the AST, maintain some state
 struct StateCheck<'a> {
     diagnostics: Vec<Diagnostic>,
+    /// Does the function have an expression or statement that reads from state?
```

Comment:
Sorry for this nitpick, but I think using an `enum Mutability { Write, Read, None }` instead a bunch of bools would make for more idiomatic code overall. The struct could then just have two fields of this enum, like `desired_mutability` and `actual_mutability`.

I mean its unrelated to this bugfix; I'll just leave it as suggestion if you feel like it

Reply:
Great idea, I've re-written it with an enum.